src/internal/connector/exchange/service_restore.go

```diff
@@ -419,257 +214,143 @@ func restoreCollection(
-			return "", err
+			return "", cache, clues.Stack(err)
 		}
 
-		folderID = ptr.Val(temp.GetId())
-
-		// Only populate the cache if we actually had to create it. Since we set
-		// newCache to false in this we'll only try to populate it once per function
-		// call even if we make a new cache.
-		if isNewCache {
-			if err := mfc.Populate(ctx, errs, rootFolderAlias); err != nil {
-				return "", clues.Wrap(err, "populating folder cache")
-			}
-
-			isNewCache = false
-		}
-
-		// NOOP if the folder is already in the cache.
-		if err = mfc.AddToCache(ctx, temp); err != nil {
-			return "", clues.Wrap(err, "adding folder to cache")
-		}
+		containerParentID = fid
 	}
 
-	return folderID, nil
+	// containerParentID now identifies the last created container,
+	// not its parent.
+	return containerParentID, cache, nil
 }
 
-// establishContactsRestoreLocation creates Contact Folders in sequence
-// and updates the container resolver appropriately. Contact Folders are
-// displayed in a flat representation. Therefore, only the root can be populated and all content
-// must be restored into the root location.
-// @param folders is the list of intended folders from root to leaf (e.g. [root ...])
-// @param isNewCache bool representation of whether Populate function needs to be run
-func establishContactsRestoreLocation(
+func getOrPopulateContainer(
 	ctx context.Context,
-	ac api.Client,
-	folders []string,
-	cfc graph.ContainerResolver,
-	user string,
+	ca containerAPI,
+	gcr graph.ContainerResolver,
+	restoreLoc *path.Builder,
+	userID, containerParentID, containerName string,
 	isNewCache bool,
 	errs *fault.Bus,
 ) (string, error) {
-	cached, ok := cfc.LocationInCache(folders[0])
+	cached, ok := gcr.LocationInCache(restoreLoc.String())
 	if ok {
 		return cached, nil
 	}
 
-	ctx = clues.Add(ctx, "is_new_cache", isNewCache)
+	c, err := ca.CreateContainer(ctx, userID, containerName, containerParentID)
+
+	// 409 handling case:
+	// attempt to fetch the container by name and add that result to the cache.
+	// This is rare, but may happen if CreateContainer() POST fails with 5xx:
+	// sometimes the backend will create the folder despite the 5xx response,
+	// leaving our local containerResolver with inconsistent state.
+	if graph.IsErrFolderExists(err) {
+		cbn, ok := ca.containerSearcher()
+		if ok {
+			c, err = cbn.GetContainerByName(ctx, userID, containerName)
+		}
+	}
 
-	temp, err := ac.Contacts().CreateContactFolder(ctx, user, folders[0])
 	if err != nil {
-		return "", err
+		return "", clues.Wrap(err, "creating restore container")
 	}
 
-	folderID := ptr.Val(temp.GetId())
+	folderID := ptr.Val(c.GetId())
 
 	if isNewCache {
-		if err := cfc.Populate(ctx, errs, folderID, folders[0]); err != nil {
-			return "", clues.Wrap(err, "populating contact cache")
+		if err := gcr.Populate(ctx, errs, folderID, ca.orRootContainer(restoreLoc.HeadElem())); err != nil {
+			return "", clues.Wrap(err, "populating container cache")
 		}
```

Comment:
we should probably actually just be populating the container resolver when we create it in the top-level function. This could cause some weird interactions down the line if we're not careful. Also advanced restore options will sort of depend on being able to do path -> container ID translations for containers that we may not have created in this loop

Reply:
Yes, I 100% agree.  I almost did that during the last batch of changes.  Held back because that's more of a logical update than was required to get the refactoring to work, and I wanted to keep the change within those bounds.  But I do plan on moving this up earlier when I start to implement advanced restores.